File: algorithms/ar-hsrt/java/HeapSort.java
public class HeapSort {

    public void sort(int arr[]){

        int n = arr.length;

        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i);
        }

        for (int i=n-1; i>0; i--){

            int temp_element = arr[0];
            arr[0] = arr[i];
            arr[i] = temp_element;

            heapify(arr, i, 0);
        }
    }

    void heapify(int arr[], int n, int i){

        int largest = i;
        int l = 2*i + 1;
        int r = 2*i + 2;

        if (l < n && arr[l] > arr[largest])
            largest = l;

        if (r < n && arr[r] > arr[largest])
            largest = r;

        if (largest != i)
        {
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;

            heapify(arr, n, largest);
        }
    }

    public static void main(String args[]){
        int input[] = {15, 2, 8, 9, 0, 1, 12, 22};
        int n = input.length;

        HeapSort ob = new HeapSort();
        ob.sort(input);

        System.out.println("Sorted array is");

        for (int i=0; i<n; ++i) {
            System.out.print(input[i] + " ");
        }
    }
}

File: algorithms/ar-hsrt/typescript/heap_sort.ts
// Heap Sort Typescript Implementation

/**
 *  Helper function to heapify the tree
 */
function heapify(collection: Array<number>, length: number, index: number): void {
    
    let rootNode: number = index; // root node location initialization
    const leftChild: number = index*2 + 1; // left child location  initialization based on root
    const rightChild: number = index*2 + 2; // right child location initialization based on root

    // if value of left child is greater than the root node
    if (leftChild < length && collection[leftChild] > collection[rootNode]) {
        rootNode = leftChild;
    }

    // if value of right child is greater than the root node
    if (rightChild < length && collection[rightChild] > collection[rootNode]) {
            rootNode = rightChild;
    }

    // if root node is not the index
    if (rootNode !== index){
        // swap the values
        const temp: number = collection[rootNode];
        collection[rootNode] = collection[index];
        collection[index] = temp;
        // Recursively sort the sub-tree
        heapify(collection, length, rootNode);
    }

}  

/**
 *  Main function to call for heap sort
 * @param collection Array of numbers
 */
function heapSort(collection: Array<number>): void {
    // Build heap by rearranging elements 
    for (let i = collection.length / 2; i >= 0; --i) {
         heapify(collection, collection.length, i);
    }
    // Iterate through the elements of heap
    for (let i = collection.length - 1; i >= 0; --i) {
        // swap current node with the leaf node
        const temp = collection[0];
        collection[0] = collection[i];
        collection[i] = temp;
        
        // heapify the sub-tree
        heapify(collection, i, 0);
    }

}

// Driver program

function main() {
  const sample: Array<number> = [12, 11, 13, 5, 6, 7];
  heapSort(sample);
}

main();
